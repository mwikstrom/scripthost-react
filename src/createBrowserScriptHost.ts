import { ScriptHost, ScriptHostOptions } from "scripthost";
import { BrowserSandbox, BrowserSandboxOptions } from "scripthost-browser";

/**
 * @public
 */
export function createBrowserScriptHost(
    options: ScriptHostOptions & Pick<BrowserSandboxOptions, "unsafe"> = {}
): ScriptHost {
    const { unsafe, ...hostOptions } = options;
    return new ScriptHost(() => new BrowserSandbox({ unsafe }), { ...hostOptions });
}
