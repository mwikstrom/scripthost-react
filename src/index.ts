/**
 * React bindings for Scripthost
 * @packageDocumentation
 */

export * from "./createBrowserScriptHost";
export * from "./ScriptHostScope";
export * from "./useObservedScript";
export * from "./useScriptInvoker";
