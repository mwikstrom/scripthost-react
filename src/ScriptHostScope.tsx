import React, { createContext, FC, ReactNode, useContext, useMemo } from "react";
import { ScriptHost } from "scripthost";
import { createBrowserScriptHost } from "./createBrowserScriptHost";

/**
 * @public
 */
export interface ScriptHostScopeProps {
    children?: ReactNode;
    host?: ScriptHost;
}

/**
 * @public
 */
export const ScriptHostScope: FC<ScriptHostScopeProps> = props => {
    const {  children } = props;
    const host = useMemo(() => props.host ?? createBrowserScriptHost(), [props.host]);
    return (
        <ScriptHostContext.Provider
            value={host}
            children={children}
        />
    );
};

/**
 * @public
 */
export const useScriptHost = (): ScriptHost => useContext(ScriptHostContext);

const ScriptHostContext = createContext<ScriptHost>(createBrowserScriptHost());
