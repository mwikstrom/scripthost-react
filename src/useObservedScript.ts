import { useLayoutEffect, useMemo, useState } from "react";
import { ScriptEvalOptions } from "scripthost";
import { ScriptValue } from "scripthost-core";
import { useScriptHost } from "./ScriptHostScope";

/**
 * @public
 */
export interface ObservedScript {
    result: ScriptValue;
    ready: boolean;
    error: Error | null;
}

/**
 * @public
 */
export type UseObservedScriptOptions = Pick<ScriptEvalOptions, "instanceId" | "timeout">;

/**
 * @public
 */
export function useObservedScript(script: string, options: UseObservedScriptOptions = {}): ObservedScript {
    const { instanceId, timeout } = options;
    const host = useScriptHost();
    const [output, setOutput] = useState(initialOutput);

    useLayoutEffect(() => {
        const onNext = (result: ScriptValue) => {
            setOutput(successOutput(result));
        };
        const onError = (err: unknown) => {
            const error = err instanceof Error ? err : new Error("Script failed");
            setOutput(errorOutput(error));
        };
        setOutput(initialOutput);
        return host.observe(script, { instanceId, timeout, onNext, onError });
    }, [host, script, instanceId, timeout]);

    return output;
}

const initialOutput: ObservedScript = Object.freeze({
    result: undefined,
    ready: false,
    error: null,
});

const successOutput = (result: ScriptValue): ObservedScript => Object.freeze({
    result,
    ready: true,
    error: null,
});

const errorOutput = (error: Error): ObservedScript => Object.freeze({
    result: undefined,
    ready: true,
    error,
});
