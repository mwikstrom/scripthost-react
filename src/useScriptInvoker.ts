import { useCallback } from "react";
import { ScriptEvalOptions } from "scripthost";
import { ScriptValue } from "scripthost-core";
import { isNonVoidScript } from "./internal/void-script";
import { useScriptHost } from "./ScriptHostScope";

/**
 * @public
 */
export type ScriptInvoker = () => Promise<ScriptValue>;

/**
 * @public
 */
export type UseScriptInvokerOptions = Pick<ScriptEvalOptions, "instanceId" | "timeout" | "vars">;

/**
 * @public
 */
export function useScriptInvoker(script: string | null, options: UseScriptInvokerOptions = {}): ScriptInvoker {
    const { instanceId, timeout, vars } = options;
    const host = useScriptHost();
    return useCallback<ScriptInvoker>(async () => {
        if (isNonVoidScript(script)) {
            await host.eval(script, { instanceId, timeout, vars });
        } else {
            return void(0);
        }
    }, [host, script, instanceId, timeout, vars]);
}