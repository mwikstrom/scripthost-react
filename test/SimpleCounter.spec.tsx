/* eslint-disable @typescript-eslint/no-non-null-assertion */
import React, { FC } from "react";
import { create, act, ReactTestRenderer } from "react-test-renderer";
import { ScriptHost } from "scripthost";
import { createBrowserScriptHost, ScriptHostScope, useObservedScript, useScriptInvoker } from "../src";

test("Simple counter", async () => {
    const host = createBrowserScriptHost({ unsafe: true });
    const element = <CounterControl host={host}/>;
    
    // Initialzing
    let renderer: ReactTestRenderer;
    act(() => void(renderer = create(element)));
    expect(renderer!.toJSON()).toMatchSnapshot("Initializing");
    
    // Expression evaluated => Value = 0
    await act(async () => {
        await host.whenIdle();
        renderer.update(element);
    });
    expect(renderer!.toJSON()).toMatchSnapshot("Value = 0");

    // Click increment button => Value = 1
    await act(async () => {
        renderer.root.findByType("button").props.onClick();
        await host.whenIdle();
        renderer.update(element);
    });
    expect(renderer!.toJSON()).toMatchSnapshot("Value = 1");

    // Click increment button => Value = 2
    await act(async () => {
        renderer.root.findByType("button").props.onClick();
        await host.whenIdle();
        renderer.update(element);
    });
    expect(renderer!.toJSON()).toMatchSnapshot("Value = 2");

    // Unmount and wait for observed script linger timeout
    jest.useFakeTimers();
    act(() => renderer!.unmount());
    jest.runAllTimers();
});

const CounterControl: FC<{ host: ScriptHost }> = ({host}) => (
    <ScriptHostScope host={host}>
        <IncrementButton/>
        <CounterValue/>
    </ScriptHostScope>
);

const CounterValue: FC = () => {
    const { result, ready } = useObservedScript("value || 0");
    if (ready) {
        return <div>Value = {String(result)}</div>;
    } else {
        return <div>Initializing</div>;
    }
};

const IncrementButton: FC = () => {
    const increment = useScriptInvoker("value = (value || 0) + 1");
    return <button onClick={increment}>Increment</button>;
};
