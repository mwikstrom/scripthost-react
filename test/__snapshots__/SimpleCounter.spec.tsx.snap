// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Simple counter: Initializing 1`] = `
Array [
  <button
    onClick={[Function]}
  >
    Increment
  </button>,
  <div>
    Initializing
  </div>,
]
`;

exports[`Simple counter: Value = 0 1`] = `
Array [
  <button
    onClick={[Function]}
  >
    Increment
  </button>,
  <div>
    Value = 
    0
  </div>,
]
`;

exports[`Simple counter: Value = 1 1`] = `
Array [
  <button
    onClick={[Function]}
  >
    Increment
  </button>,
  <div>
    Value = 
    1
  </div>,
]
`;

exports[`Simple counter: Value = 2 1`] = `
Array [
  <button
    onClick={[Function]}
  >
    Increment
  </button>,
  <div>
    Value = 
    2
  </div>,
]
`;
